### DISCUSSION TITLE
Is GpioLcdDisplay available in Pi4J v2?

### DISCUSSION THREAD

**Discussion by crdavis-edu** [Pi4J/pi4j#94] [Q&A] [answered]
I need to create demo code to illustrate to my students how characters can be displayed on a LCD using Pi4J.
Pi4J v1 has the GpioLcdDisplay class that simplifies this task. Is GpioLcdDisplay available in Pi4J v2?
If it isn't, can someone point me to code that can be use to control a LCD using Pi4J v2.

Thanks in advance for responding.

**Comment by FDelporte** [comment, marked as answer]
It depends on which type of LCD display you are using. Personally, I lost my love for them when using one in which you had to control all the pins yourself ;-) But after switching to an I2C type for an Arduino project I feel the love returning...

E.g. like this one https://www.parallax.com/product/16x2-i2c-lcd-display-module-with-blue-backlight/

At this moment, students of FHNW University are creating example applications with the CrowPi and LCD is one part of that project. Maybe they can already share some example code here? Will ask...

**Reply by crdavis-edu** [reply]
Thanks Frank for your prompt response. I am using  LCD1602 display screen in conjunction with  I2C LCD 1602 module; these devices are part of Freenove Ultimate Starter Kit for Raspberry Pi (https://www.amazon.ca/Freenove-Raspberry-Processing-Tutorials-Components/dp/B06W54L7B5).
Thanks in advance for enquiring about example code from students of FHNW University.

Best regards.

**Reply by ppmathis** [reply]
@crdavis-edu Hi there! I've taken a look at the Freenove kit you've linked and while it also uses the `LCD1602` component, it seems to use a slightly different I/O expander for connecting the LCD itself. This means that our current implementation isn't an exact fit, but you should get it working by adjusting it for the I/O expander used in your kit.

While we are still working on getting our code from another repository into the official Pi4J CrowPi examples repository, I can already provide you with a couple links where you can access our current work:

- [LcdDisplayApp.java](https://github.com/Pi4J/pi4j-example-crowpi/blob/main/src/main/java/ch/fhnw/crowpi/applications/LcdDisplayApp.java) contains an example application which uses our LCD component and demonstrates various abilities of the LCD1602 display
- [LcdDisplayComponent.java](https://github.com/Pi4J/pi4j-example-crowpi/blob/main/src/main/java/ch/fhnw/crowpi/components/LcdDisplayComponent.java) is the actual component class for interacting with the `LCD1602` component. Rather than directly messing with any GPIO pins though, it instantiates `MCP23008` which is the I²C I/O expander used by the CrowPi.
- [MCP23008.java](https://github.com/Pi4J/pi4j-example-crowpi/blob/main/src/main/java/ch/fhnw/crowpi/components/internal/MCP23008.java) is the class for the I/O expander mentioned before. It uses I²C to talk with the expander and control the individual pins of the display. This is also where Pi4J v2 comes into play :smile: 

In your case you should be able to create your own I/O expander class for the `PCF8574` chip used in your kit and hook that into the `LcdDisplayComponent` class to get everything up and running. As the I/O expander is just a really small part of the puzzle, this should hopefully not take too much time to achieve. Looking at [the Python implementation](https://github.com/Freenove/Freenove_LCD1602_Starter_Kit_for_Raspberry_Pi/blob/master/Code/Python_Code/11.1.1_I2CLCD1602/PCF8574.py) for your I/O expander as a reference might also be helpful.

Feel free to ask if any further questions about the shared code come up on your side. I hope this helps!

P.S.: Previous students also once worked with this LCD and published [this code](https://github.com/FHNW-IP5-IP6/GPIOSimulator/blob/master/src/main/java/fhnwgpio/components/I2CLCDComponent.java) which also uses I²C as an I/O expander and should work ootb for the Freenove kit, but is only based on Pi4J v1.2. This might be useful as some additional reference.

**Reply by crdavis-edu** [reply]
Thanks much Pascal for your detailed response. Your pointers are very helpful. I will take a more detailed look at the code you provided; hopefully, we will be able to code the necessary tweak to get us up and running.

Best regards,

**Reply by FDelporte** [reply]
+1 for the clear description 
+1 for the clear separation in the code between app - component - I/O expander
+∞ for sharing here :-)

**Reply by crdavis-edu** [reply]
By-the-way Frank, I bought your "Getting started with Java on Raspberry Pi" book last year and I did a re-read a couple weeks ago.  As a fellow author of tech books, I know that writing books is a very time consuming task. Thanks for taking the time to write and publish a very useful book.

**Reply by FDelporte** [reply]
Thanks for your kind words @crdavis-edu! Was indeed a big job and required a lot of time, but had fun at the same time ;-)
